File: LeetCode/Problems/Python/238. Product of Array Except Self.py
#Time: O(n)
#Space: O(n)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #brute force:
        #iterate and do the calcuation from any index
        #lead to O(n^2)
        
        #method 1, preprocess, and divide by every index value
        #Time: O(n)
        
        #method 2, preprocess from both side, product of all
        #at every index we pick the left and right from each list and multiply them
        n = len(nums)
        forward = [1] * n 
        backward = [1] * n
        forward[0] = 1
        backward[-1] = 1
        for i in range(1,len(nums)):
            forward[i] = forward[i-1] * nums[i-1]
            backward[~i] = backward[~i+1] * nums[~i+1]
        print(forward,backward)
        
        ans = [1] * n
        for i in range(len(nums)):
            ans[i] = forward[i] * backward[i]
        return ans

#My optimization
#Time: O(n)
#Space: O(1)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #brute force:
        #iterate and do the calcuation from any index
        #lead to O(n^2)
        
        #method 1, preprocess, and divide by every index value
        #Time: O(n)
        
        #method 2, preprocess from both side, product of all
        #at every index we pick the left and right from each list and multiply them
        n = len(nums)
        forward = 1
        backward = 1
        ans = [1] * n
        for i in range(1,len(nums)):
            
            ans[i] *= forward * nums[i-1]
            ans[~i] *= backward * nums[~i+1]
            forward = forward * nums[i-1]
            backward = backward * nums[~i+1]
            
        print(forward,backward)

        return ans

#two pass clear answer
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #brute force:
        #iterate and do the calcuation from any index
        #lead to O(n^2)
        
        #method 1, preprocess, and divide by every index value
        #Time: O(n)
        
        #method 2, preprocess from both side, product of all
        #at every index we pick the left and right from each list and multiply them
        n = len(nums)


        ans = [1] * n
        for i in range(1,len(nums)):
            #forward pass
            ans[i]*= ans[i-1] * nums[i-1]
        right = 1
        for i in range(len(nums)-1,-1,-1):
            #backward pass
            ans[i] *= right
            right *= nums[i]
            
        #print(ans)
        return ans




#second try
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        n = len(nums)


        ans = [1] * n
        for i in range(1,len(nums)):
            #forward pass
            ans[i]*= ans[i-1] * nums[i-1]
        right = 1
        for i in range(len(nums)-1,-1,-1):
            ans[i] *= right
            right *= nums[i]

        return ans

File: LeetCode/Problems/C++/238. Product of Array Except Self.cpp
//Time: O(n)
//Space: O(1)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        if(!nums.size()) return {};

        vector<int> result;
        result.resize(nums.size(),1);
        int right_product = 1;

        for(int i=1;i<nums.size();i++)
            result[i] = result[i-1] * nums[i-1];
        
        
        
        for(int i=nums.size()-1;i>=0;i--){
            result[i] *= right_product;
            right_product*=nums[i];
        }
        return result;
    }
};

//Time: O(n)
//Space: O(n)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        if(!nums.size()) return {};
        
        vector<int> front_prefix;
        vector<int> back_prefix;
        vector<int> result;
        int prefix_product = 1;
        for(int i=0;i<nums.size();i++){
            front_prefix.push_back(prefix_product);
            prefix_product*= nums[i];
        }
        prefix_product = 1;
        for(int i=nums.size()-1;i>=0;i--){
            back_prefix.push_back(prefix_product);
            prefix_product*= nums[i];
        }
        
        for(int i=0;i<nums.size();i++){
            result.push_back(front_prefix[i]*back_prefix[nums.size()-1-i]);
        }
        
        return result;
    }
};